dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1507 @@
+            {
+                UpdateSizeBasedTemplateSettings();
+                // Reset the currentEffectivePlacementMode so that the beak will be updated for the new size as well.
+                m_currentEffectivePlacementMode = winrt::TeachingTipPlacementMode::Auto;
+                TeachingTipTestHooks::NotifyEffectivePlacementChanged(*this);
+                if (IsOpen())
+                {
+                    PositionPopup();
+                }
+                if (m_expandAnimation)
+                {
+                    m_expandAnimation.get().SetScalarParameter(L"Width", static_cast<float>(m_beakOcclusionGrid.get().ActualWidth()));
+                    m_expandAnimation.get().SetScalarParameter(L"Height", static_cast<float>(m_beakOcclusionGrid.get().ActualHeight()));
+                }
+                if (m_contractAnimation)
+                {
+                    m_contractAnimation.get().SetScalarParameter(L"Width", static_cast<float>(m_beakOcclusionGrid.get().ActualWidth()));
+                    m_contractAnimation.get().SetScalarParameter(L"Height", static_cast<float>(m_beakOcclusionGrid.get().ActualHeight()));
+                }
+            }
+        });
+    }
+    if (m_closeButton)
+    {
+        m_closeButton.get().Click({ this, &TeachingTip::OnCloseButtonClicked });
+    }
+    if (m_alternateCloseButton)
+    {
+        m_alternateCloseButton.get().Click({ this, &TeachingTip::OnCloseButtonClicked });
+    }
+
+    if (m_actionButton)
+    {
+        m_actionButton.get().Click({ this, &TeachingTip::OnActionButtonClicked });
+    }
+    UpdateButtonsState();
+    OnIconSourceChanged();
+    EstablishShadows();
+
+    if (m_startAnimationInOnApplyTemplate)
+    {
+        StartExpandToOpen();
+        m_startAnimationInOnApplyTemplate = false;
+    }
+}
+
+void TeachingTip::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+
+    if (property == s_IsOpenProperty)
+    {
+        OnIsOpenChanged();
+    }
+    else if (property == s_ActionButtonTextProperty ||
+        property == s_CloseButtonKindProperty  ||
+        property == s_CloseButtonTextProperty)
+    {
+        UpdateButtonsState();
+    }
+    else if (property == s_TargetOffsetProperty)
+    {
+        OnTargetOffsetChanged();
+    }
+    else if (property == s_IsLightDismissEnabledProperty)
+    {
+        OnIsLightDismissEnabledChanged();
+    }
+    else if (property == s_PlacementProperty)
+    {
+        if (IsOpen())
+        {
+            PositionPopup();
+        }
+    }
+    else if (property == s_BleedingImagePlacementProperty)
+    {
+        OnBleedingImagePlacementChanged();
+    }
+    else if (property == s_IconSourceProperty)
+    {
+        OnIconSourceChanged();
+    }
+}
+
+void TeachingTip::OnContentChanged(const winrt::IInspectable& oldContent, const winrt::IInspectable& newContent)
+{
+    if (Content())
```

Comment:
Nit: you can check if(newContent) instead of having to retrieve the Content property (which requires a QI)